cmd/util/signal.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+var onlyOneSignalHandler = make(chan struct{})
+var errorExitCodeChannel = make(chan int, 1)
+
+// SetupExitHandler:
+// A stop channel is returned which is closed on receiving a shutdown signal (SIGTERM
+// or SIGINT). If a second signal is caught, the program is terminated directly with
+// exit code 130.
+// The function also returns an exit function, this function calls os.Exit(...) and is
+// the last function that will be executed by the program. The exit function will try
+// to receive an exit code from the errorExitCodeChannel. If no exit code can be
+// retrieved, 0 will be used as exit code.
+// The errorExitCodeChannel receives exit codes when GracefullyExit is called or when
+// a shutdown signal is received (only if isLongRunning is false).
+func SetupExitHandler(isLongRunning bool) (<-chan struct{}, func()) {
```

Comment:
suggestion: a quick additional comment, as a side-note to the much-more-important discussion about exit codes.

I'm not a huge fan of boolean parameters like this, because they don't convey any meaning when you see them at a call site:

```go
SetupExitHandler(true) //  what does true mean?
```
Bool parameters are also inflexible if we decide there's a third option down the road; we end up having to create a new function or break the API.

Instead, how about having a new type? It's a little more code, but I think it's a little more self documenting:

```go
package main

import "fmt"

// ExitBehavior controls how the program should be terminated
// in response to a shutdown signal.
type ExitBehavior int

const (
	// AlwaysErrCode indicates the exit code of the program should always be nonzero
	// and should correspond to the numeric value of the signal that was received.
	AlwaysErrCode ExitBehavior = iota

	// GracefulShutdown treats a shutdown signal as a request to exit gracefully, terminating
	// goroutines and returning an exit code of 0 if there are no errors during shutdown.
	GracefulShutdown ExitBehavior = iota
)

func SetupExitHandler(behavior ExitBehavior) (<-chan struct{}, func()) {
	// impl
}

func main() {
	SetupExitHandler(AlwaysErrCode)
	SetupExitHandler(GracefulShutdown)
}
```

What do you think?